sqlalchemy-stubs/sql/functions.pyi

```diff
@@ -1,250 +1,308 @@
+from .elements import ClauseElement
+from .elements import ClauseList
+from .elements import ColumnElement
+from .elements import FunctionFilter
+from .elements import Grouping
+from .elements import NamedColumn
+from .elements import Over
+from .elements import TableValuedColumn
+from .elements import WithinGroup
+from .schema import Sequence
+from .selectable import FromClause
+from .selectable import Select
+from .selectable import TableValuedAlias
+from .visitors import TraversibleType
+
+_TE = TypeVar("_TE", bound=type_api.TypeEngine[Any])
+_FE = TypeVar("_FE", bound=FunctionElement[Any])
 
 def register_function(
-    identifier: Any, fn: Any, package: str = ...
+    identifier: str, fn: Any, package: str = ...
 ) -> None: ...
 
-class FunctionElement(Executable, ColumnElement, FromClause, Generative):
+class FunctionElement(  # type: ignore[misc]
+    Executable, ColumnElement[_TE], FromClause, Generative, Generic[_TE]
+):
     packagenames: Any = ...
     clause_expr: Any = ...
     def __init__(self, *clauses: Any, **kwargs: Any) -> None: ...
-    def scalar_table_valued(self, name: Any, type_: Optional[Any] = ...): ...
-    def table_valued(self, *expr: Any, **kw: Any): ...
-    def column_valued(self, name: Optional[Any] = ...): ...
+    @overload
+    def scalar_table_valued(
+        self, name: Any, type_: None = ...
+    ) -> ScalarFunctionColumn[sqltypes.NullType]: ...
+    @overload
+    def scalar_table_valued(
+        self, name: Any, type_: Union[_TE, Type[_TE]]
+    ) -> ScalarFunctionColumn[_TE]: ...
+    def table_valued(self, *expr: Any, **kw: Any) -> TableValuedAlias: ...  # type: ignore[override]
+    def column_valued(
+        self, name: Optional[Any] = ...
+    ) -> TableValuedColumn[Any]: ...
     @property
-    def columns(self): ...
-    def clauses(self): ...
+    def columns(self) -> ColumnCollection[Any]: ...  # type: ignore[override]
+    @HasMemoized.memoized_attribute
+    def clauses(self) -> ClauseList[Any]: ...
     def over(
         self,
-        partition_by: Optional[Any] = ...,
-        order_by: Optional[Any] = ...,
+        partition_by: Optional[ClauseElement] = ...,
+        order_by: Optional[ClauseElement] = ...,
         rows: Optional[Any] = ...,
         range_: Optional[Any] = ...,
-    ): ...
-    def within_group(self, *order_by: Any): ...
-    def filter(self, *criterion: Any): ...
-    def as_comparison(self, left_index: Any, right_index: Any): ...
-    def within_group_type(self, within_group: Any) -> None: ...
-    def alias(self, name: Optional[Any] = ...): ...
-    def select(self): ...
-    def scalar(self): ...
-    def execute(self): ...
-    def self_group(self, against: Optional[Any] = ...): ...
-
-class FunctionAsBinary(BinaryExpression):
-    sql_function: Any = ...
-    left_index: Any = ...
-    right_index: Any = ...
+    ) -> Over[_TE]: ...
+    def within_group(self, *order_by: Any) -> WithinGroup[_TE]: ...
+    @overload
+    def filter(self) -> _FE: ...
```

Comment:
should this be `filter(self: _FE) -> _FE`?

Reply:
It should. I'll push an update for it.